OrbitCore/SymbolsManager.cpp

```diff
@@ -0,0 +1,150 @@
+#include "SymbolsManager.h"
+
+#include "Capture.h"
+#include "ConnectionManager.h"
+#include "CoreApp.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitModule.h"
+#include "SymbolHelper.h"
+#include "TransactionManager.h"
+
+namespace orbit {
+
+SymbolsManager& SymbolsManager::Get() {
+  static SymbolsManager symbols_manager;
+  return symbols_manager;
+}
+
+void SymbolsManager::Init() {
+  auto on_response = [this](const Message& msg) { HandleResponse(msg); };
+  auto on_request = [this](const Message& msg) { HandleRequest(msg); };
+  TransactionManager::Get().RegisterTransactionHandler(
+      {on_request, on_response, Msg_DebugSymbols, "Debug Symbols"});
+}
+
+void SymbolsManager::LoadSymbols(std::shared_ptr<Session> session,
+                                 std::shared_ptr<Process> process) {
+  session_ = session;
+  std::vector<std::shared_ptr<Module>> modules;
+  for (auto& pair : session->m_Modules) {
+    const std::string& file_name = Path::GetFileName(pair.first);
+    std::shared_ptr<Module> module = process->GetModuleFromName(file_name);
+    if (module != nullptr) {
+      modules.emplace_back(module);
+    }
+  }
+  LoadSymbols(modules, process);
+}
+
+void SymbolsManager::LoadSymbols(
+    const std::vector<std::shared_ptr<Module>>& modules,
+    std::shared_ptr<Process> process) {
+  if (modules.empty()) {
+    ERROR("No module to load, cancelling.");
+    return;
+  }
+
+  if (request_in_flight_) {
+    ERROR("Module request already in flight, cancelling.");
+    return;
+  }
+
+  request_in_flight_ = true;
```

Comment:
This was written from the perspective of all loading requests being issued from the main thread, while responses would be handled on a different thread.  But this could indeed made better.